File: macos/Sources/Voicekey/Core/StreamingTranscriber.swift
//
//  StreamingTranscriber.swift
//  Deepgram リアルタイム文字起こし（WebSocket ストリーミング）
//
//  録音中の 16kHz PCM チャンクを WebSocket で逐次送り、暫定（interim）と
//  確定（is_final）のテキストをリアルタイムに受け取る。HUD のライブ字幕に使う。
//  ベンチマークで Deepgram nova-3 が「離した瞬間にほぼ確定（60〜170ms）」かつ
//  最高精度だったため、ストリーミングの既定バックエンドとして採用している。
//
//  外部依存を増やさないため URLSessionWebSocketTask（標準 API）を使う。
//  送信は audio スレッド、受信は URLSession のキュー、開始/終了はメインから
//  呼ばれるため、共有状態は NSLock で保護する。
//

import Foundation
import os.log

private let log = Logger(subsystem: "com.voicekey.app", category: "stream")

/// Deepgram WebSocket による逐次文字起こしセッション（1 録音 = 1 インスタンス）
final class StreamingTranscriber: @unchecked Sendable {

    /// 現在の全文（確定 + 暫定）の更新通知。HUD 用にメインへホップして使う想定
    var onInterim: ((String) -> Void)?

    private let model: String
    private let language: String
    private let session: URLSession
    private var task: URLSessionWebSocketTask?

    private let lock = NSLock()
    /// 確定済みセグメント（is_final=true の順次追記）
    private var finals: [String] = []
    /// 直近の暫定セグメント（次の確定で置き換わる）
    private var interim: String = ""
    /// 接続確立前に届いた PCM の退避（接続時に順序を保ってフラッシュする）。
    /// 製品版（ログイン）は短命 JWT 取得が非同期なので、接続前に send() が来うる。
    private var pending: [Data] = []
    /// finish/cancel が接続確立前に呼ばれた（接続完了後に即破棄させる）
    private var cancelled = false
    /// finish() を解決する継続。二重 resume を防ぐため lock 下で nil 化する
    private var finishContinuation: CheckedContinuation<Void, Never>?
    /// 接続が閉じた / 確定し終えたか
    private var done = false
    /// 録音開始遅延の計測用（生成時刻 ≒ 録音開始）。トークン往復 / 最初の文字までの累計を
    /// ログし、「始まりが遅い」の主因（サーバー往復か WS ハンドシェイクか）を裏取りする
    private let createdAt = Date()
    private var firstResultLogged = false
    /// 無料体験の保留 ID（段階1 hold 経路のみ非null）。文字起こし成功後に消費を確定するのに使う。
    private var ephemeralJti: String?
    /// 段階3: 再利用トークン（free）＝録音成功ごとに jti なし confirm で 1 消費すべきか。
    private var ephemeralMeter = false

    init(model: String, language: String) {
        self.model = model
        self.language = language
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: cfg)
    }

    /// Deepgram の言語パラメータ（REST 側 Transcriber と同じ規則）。
    /// nova-3 は日本語の単言語指定に未対応のため多言語モードを使う。
    private var streamLanguage: String {
        if model.hasPrefix("nova-3") { return "multi" }
        return language.isEmpty ? "ja" : language
    }

    /// WebSocket を開いて受信ループを開始する。
    ///
    /// 製品版（ログイン済み）はサーバーから短命 JWT を取得して `Bearer` で接続する。
    /// この取得は非同期なので接続確立前に send() が来うる（→ pending に退避）。
    /// 未ログインは従来どおり埋め込み/設定キーを `Token` で使う（並存ガード）。
    /// - Returns: キー未設定かつ未ログインなど開始できない場合 false（呼び出し側は REST にフォールバック）
    func start() -> Bool {
        let loggedIn = BackendClient.isLoggedIn
        // 配布版（製品版ビルド）はログイン必須＝未ログインでは埋め込みキーを使わない。
        // その場合 false を返し、ゲート済みの REST 経路に委ねてログイン要求エラーを表示させる。
        let key = EmbeddedKeys.isDist ? nil : Keychain.apiKey(for: .deepgram)
        guard loggedIn || key != nil else {
            log.error("Deepgram: 未ログイン（配布版はログイン必須）またはキー未設定のためストリーミングを開始できません")
            return false
        }

        if loggedIn {
            // 製品版: サーバーから短命 JWT を取得してから Bearer で接続する
            Task { [weak self] in
                guard let self else { return }
                do {
                    let t0 = Date()
                    let tok = try await BackendClient.fetchEphemeralToken()
                    // 録音成功後に消費を確定するための情報を控える。段階1=保留 ID(jti)、
                    // 段階3=再利用トークンの回数計測フラグ(meter)。どちらも録音成立時のみ送る。
                    self.lock.lock(); self.ephemeralJti = tok.jti; self.ephemeralMeter = tok.meter; self.lock.unlock()
                    // 「始まりが遅い」の主因切り分け用ログ（トークン往復 ms）
                    log.info("Deepgram 短命トークン取得 \(Int(Date().timeIntervalSince(t0) * 1000), privacy: .public)ms")
                    self.connect(auth: "Bearer \(tok.token)")
                } catch {
                    log.error("Deepgram 短命トークンの取得に失敗: \(error.localizedDescription)")
                    self.resolveFinish()  // finish() を解決し REST フォールバックに委ねる
                }
            }
        } else {
            // 未ログイン: 従来の Token 直叩き（key は guard で非 nil 確定）
            connect(auth: "Token \(key!)")
        }
        return true
    }

    /// 指定の Authorization ヘッダで WebSocket を開き、退避 PCM をフラッシュして受信を開始する。
    /// 接続確立前に finish/cancel されていたら（cancelled/done）即破棄する。
    private func connect(auth: String) {
        var comps = URLComponents(string: "wss://api.deepgram.com/v1/listen")!
        comps.queryItems = [
            URLQueryItem(name: "model", value: model),
            URLQueryItem(name: "language", value: streamLanguage),
            URLQueryItem(name: "encoding", value: "linear16"),
            URLQueryItem(name: "sample_rate", value: "16000"),
            URLQueryItem(name: "channels", value: "1"),
            URLQueryItem(name: "interim_results", value: "true"),
            URLQueryItem(name: "punctuate", value: "true"),
            URLQueryItem(name: "smart_format", value: "true"),
        ]
        var request = URLRequest(url: comps.url!)
        request.setValue(auth, forHTTPHeaderField: "Authorization")

        let task = session.webSocketTask(with: request)

        // 退避済み PCM を順序保証のためロック下で取り出してから task を公開する
        lock.lock()
        if cancelled || done {
            lock.unlock()
            task.cancel(with: .normalClosure, reason: nil)
            resolveFinish()
            return
        }
        self.task = task
        let buffered = pending
        pending = []
        lock.unlock()

        task.resume()
        for chunk in buffered {
            task.send(.data(chunk)) { error in
                if let error { log.debug("退避 PCM 送信エラー: \(error.localizedDescription)") }
            }
        }
        receiveLoop()
        log.info("Deepgram ストリーミング開始 (model=\(self.model, privacy: .public), lang=\(self.streamLanguage, privacy: .public))")
    }

    /// 16kHz モノラル Float32 チャンクを送信する（audio スレッドから呼ばれる）。
    /// 接続確立前のチャンクは pending に退避し、connect() がフラッシュする。
    func send(_ samples: [Float]) {
        guard !samples.isEmpty else { return }
        // Float32[-1,1] → Int16 LE 生バイト（範囲外はクリップして轟音化を防ぐ）
        var pcm = Data(capacity: samples.count * 2)
        for s in samples {
            let clipped = max(-1.0, min(1.0, s))
            var v = Int16(clipped * 32767).littleEndian
            withUnsafeBytes(of: &v) { pcm.append(contentsOf: $0) }
        }
        lock.lock()
        if cancelled || done {
            lock.unlock()
            return
        }
        guard let task else {
            // まだ接続前 → 退避（接続確立時に connect がフラッシュする）
            pending.append(pcm)
            lock.unlock()
            return
        }
        lock.unlock()
        task.send(.data(pcm)) { error in
            if let error { log.debug("送信エラー: \(error.localizedDescription)") }
        }
    }

    /// 送信を打ち切り、確定テキストを返す（ホットキーを離したときに呼ぶ）。
    /// CloseStream を送ると Deepgram は残りを確定して Metadata を返し接続を閉じる。
    /// 取りこぼし対策で最大 3 秒だけ待つ。
    func finish() async -> String {
        // CloseStream で残りバッファの確定を促す（接続前なら task は nil で no-op）
        currentTask()?.send(.string("{\"type\":\"CloseStream\"}")) { _ in }

        await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
            lock.lock()
            if done {
                lock.unlock()
                cont.resume()
                return
            }
            finishContinuation = cont
            lock.unlock()
            // 確定が来ない場合の保険（接続ハング対策）
            DispatchQueue.global().asyncAfter(deadline: .now() + 3) { [weak self] in
                self?.resolveFinish()
            }
        }
        // 接続が JWT 取得中などで未確立でも、後から connect が走らないよう cancelled を立てる
        markCancelled()?.cancel(with: .normalClosure, reason: nil)
        // 録音のたびに生成するセッションは明示的に破棄する（放置すると漸増リーク）
        session.finishTasksAndInvalidate()
        let text = TextNormalize.stripCJKSpaces(currentText())
        // 文字起こしが成立したら無料体験の消費を確定する（ベストエフォート・非ブロッキング）。
        // 空文字（無音/接続失敗で REST フォールバック）のときは確定しない＝保留は TTL で戻る。
        if !text.isEmpty {
            lock.lock(); let jti = ephemeralJti; let meter = ephemeralMeter; ephemeralJti = nil; ephemeralMeter = false; lock.unlock()
            if let jti {
                Task { await BackendClient.confirmUsage(jti: jti) }  // 段階1: 保留を確定
            } else if meter {
                Task { await BackendClient.confirmUsageCount() }     // 段階3: 再利用トークンの回数を +1
            }
        }
        return text
    }

    /// lock 下で現在の task を読む（async コンテキストから直接 lock しないため）
    private func currentTask() -> URLSessionWebSocketTask? {
        lock.lock(); defer { lock.unlock() }
        return task
    }

    /// finish/cancel 確定: cancelled を立て、その時点の task を返す（async から呼ぶ用）
    @discardableResult
    private func markCancelled() -> URLSessionWebSocketTask? {
        lock.lock(); defer { lock.unlock() }
        cancelled = true
        return task
    }

    /// 結果を使わずに接続を破棄する（録音破棄時など）
    func cancel() {
        lock.lock()
        cancelled = true
        let t = task
        lock.unlock()
        t?.cancel(with: .normalClosure, reason: nil)
        session.finishTasksAndInvalidate()
        resolveFinish()
    }

    /// 確定 + 暫定を結合した現在の全文（生・スペース除去前）
    private func currentText() -> String {
        lock.lock(); defer { lock.unlock() }
        return (finals + (interim.isEmpty ? [] : [interim])).joined()
    }

    // MARK: - 受信

    private func receiveLoop() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure:
                // 切断・エラーは finish() を解決して REST 側の判断に委ねる
                self.resolveFinish()
            case .success(let message):
                switch message {
                case .string(let text): self.handle(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) { self.handle(text) }
                @unknown default: break
                }
                self.receiveLoop()
            }
        }
    }

    /// Deepgram のストリーミング応答（必要なフィールドのみ）
    private struct DGMessage: Decodable {
        let type: String?
        let is_final: Bool?
        let channel: Channel?
        struct Channel: Decodable { let alternatives: [Alternative] }
        struct Alternative: Decodable { let transcript: String }
    }

    private func handle(_ json: String) {
        guard let data = json.data(using: .utf8),
              let msg = try? JSONDecoder().decode(DGMessage.self, from: data) else { return }

        // Metadata は CloseStream 後の終端通知 → 確定完了とみなす
        if msg.type == "Metadata" {
            resolveFinish()
            return
        }

        guard let transcript = msg.channel?.alternatives.first?.transcript else { return }
        let trimmed = transcript.trimmingCharacters(in: .whitespacesAndNewlines)

        // 録音開始から「最初の文字が出るまで」の体感遅延を 1 度だけログ（内訳の合計確認用）
        if !trimmed.isEmpty {
            lock.lock()
            let alreadyLogged = firstResultLogged
            firstResultLogged = true
            lock.unlock()
            if !alreadyLogged {
                log.info("Deepgram 最初の文字まで \(Int(Date().timeIntervalSince(self.createdAt) * 1000), privacy: .public)ms（録音開始から）")
            }
        }

        lock.lock()
        if msg.is_final == true {
            if !trimmed.isEmpty { finals.append(trimmed) }
            interim = ""
        } else {
            interim = trimmed
        }
        let snapshot = (finals + (interim.isEmpty ? [] : [interim])).joined()
        lock.unlock()

        // HUD へは日本語スペースを除去した見た目で渡す
        onInterim?(TextNormalize.stripCJKSpaces(snapshot))
    }

    /// finish() の継続を一度だけ解決する（接続終了 / Metadata / タイムアウトのいずれか）
    private func resolveFinish() {
        lock.lock()
        let cont = finishContinuation
        finishContinuation = nil
        done = true
        lock.unlock()
        cont?.resume()
    }

}

File: src/core/streaming_transcriber.py
"""
Deepgram リアルタイム文字起こしモジュール（WebSocket ストリーミング）

録音中の 16kHz PCM チャンクを WebSocket で逐次送り、暫定（interim）と確定
（is_final）のテキストをリアルタイムに受け取る。HUD のライブ字幕に使う。
ベンチマーク（2026-06-10）で Deepgram nova-3 が「離した瞬間にほぼ確定
（94〜109ms）」かつ最高精度だったため、ストリーミングの既定バックエンドに採用。

Mac 版 StreamingTranscriber.swift（URLSessionWebSocketTask）の Python 移植。
アプリ全体がスレッドベース設計（asyncio 不使用）のため、ここでも websockets の
同期クライアント（websockets.sync.client）を使い、受信は専用スレッドで回す:

- 送信（send）: audio コールバックスレッドから呼ばれる
- 受信ループ（_run）: 接続確立 → 暫定/確定の受信 → 終端検出
- 開始/終了（start/finish）: アプリのワーカースレッドから呼ばれる
共有状態は _lock（threading.Lock）で保護する。

websockets が未導入の環境では start() が False を返し、呼び出し側は REST
文字起こしへ自動フォールバックする（keyring 不在時と同じ縮退方針）。
"""

import json
import os
import queue
import threading
import time
from typing import Callable, List, Optional
from urllib.parse import urlencode

import numpy as np
import numpy.typing as npt

from . import backend_client
from .text_utils import strip_cjk_spaces
from ..utils import secrets
from ..utils.logger import get_logger

logger = get_logger(__name__)

# 接続確立のタイムアウト（秒）
_OPEN_TIMEOUT: float = 5.0
# finish() で確定を待つ最大時間（秒）。接続ハング時の保険
_FINISH_TIMEOUT: float = 3.0
# 受信ループのポーリング間隔（秒）。closed フラグを定期確認するため
_RECV_POLL_SEC: float = 0.5
# finish() が接続確立前に呼ばれた時、接続を待つ猶予（秒）。
# 短い発話（接続確立 ~0.2s より短いキータップ）でも取りこぼさないため
_CONNECT_GRACE_SEC: float = 1.0

# 送信キューの固定上限（チャンク数）。回線が遅く送信が音声生成に追いつかなくなったら
# 音声を黙って捨てず overflow を streaming 失敗として扱い、REST フォールバックへ移す。
# 16kHz・1チャンク数十ms 換算で 512 ≒ 十数秒ぶんの退避（短い発話なら全段キャッシュ可）
_SEND_QUEUE_MAX: int = 512

# 送信キューに積む「CloseStream を送れ」マーカー（finish が退避済み音声の後に積む）
_CLOSE_STREAM = object()


def _float32_to_pcm16le(samples: npt.NDArray[np.float32]) -> bytes:
    """
    float32[-1,1] のモノラル音声を 16bit リトルエンディアン PCM のバイト列に変換する。

    範囲外サンプルはクリップして int16 ラップアラウンドによる轟音化を防ぐ。
    """
    clipped = np.clip(samples, -1.0, 1.0)
    # '<i2' で明示的にリトルエンディアン int16（Deepgram の encoding=linear16 用）
    return (clipped * 32767).astype("<i2").tobytes()


class StreamingTranscriber:
    """
    Deepgram WebSocket による逐次文字起こしセッション（1 録音 = 1 インスタンス）。

    Attributes:
        on_interim: 全文（確定 + 暫定）の更新通知コールバック。HUD 用。
                    受信スレッドから呼ばれるため、UI 反映側はメインスレッドへホップすること。
    """

    def __init__(
        self,
        model: str,
        language: str = "ja",
        on_interim: Optional[Callable[[str], None]] = None,
    ) -> None:
        """
        StreamingTranscriber を初期化する（ネットワークアクセスはしない）。

        Args:
            model: Deepgram モデル名（nova-3 / nova-2）
            language: 言語コード（nova-3 では内部的に multi に変換される）
            on_interim: 暫定テキスト更新コールバック
        """
        self.model = model
        self.language = language
        self.on_interim = on_interim

        self._lock = threading.Lock()
        self._ws = None                       # 接続確立後にセット（websockets ClientConnection）
        self._finals: List[str] = []          # 確定済みセグメント
        self._interim: str = ""               # 直近の暫定セグメント
        # PCM 送信は固定上限キュー経由で専用 sender スレッドが行う（audio コールバック内で
        # ネットワーク I/O もロック取得もしないため）。接続前に届いた PCM もここに積まれ、
        # 接続確立後に sender が FIFO で送るので順序が保たれる
        self._send_q: "queue.Queue" = queue.Queue(maxsize=_SEND_QUEUE_MAX)
        self._cancelled = False               # finish/cancel が接続前に呼ばれた
        self._closed = False                  # 受信ループ停止フラグ
        self._failed = False                  # キュー溢れ/送信失敗 → REST フォールバックへ委ねる
        self._done_event = threading.Event()  # 確定完了/切断の通知
        self._thread: Optional[threading.Thread] = None
        self._sender: Optional[threading.Thread] = None  # 送信専用スレッド

    @property
    def _stream_language(self) -> str:
        """
        Deepgram の言語パラメータ（REST 側 DeepgramTranscriber と同じ規則）。

        nova-3 は日本語の単言語指定に未対応のため多言語モード（multi）を使う。
        """
        if self.model.startswith("nova-3"):
            return "multi"
        return self.language or "ja"

    def _resolve_api_key(self) -> Optional[str]:
        """Deepgram の API キーを Keychain → 環境変数の順で解決する。"""
        return secrets.get_api_key(secrets.SERVICE_DEEPGRAM) or os.environ.get("DEEPGRAM_API_KEY")

    def start(self) -> bool:
        """
        WebSocket 受信スレッドを起動する（即座に返る）。

        実際の接続確立は受信スレッド側で行うため、ここではブロックしない。
        接続前に届いた PCM は send() が送信キューへ積み、接続確立後に sender スレッドが
        FIFO で送出するので順序は保たれる。

        Returns:
            開始できれば True。キー未設定かつ未ログイン / websockets 未導入なら False
            （呼び出し側は REST にフォールバックする）
        """
        # 製品版（ログイン済み）はサーバーから短命 JWT を取得して接続する。
        # 開発ビルドは従来どおり埋め込み/設定キーを使う（並存ガード）。
        # 配布版は埋め込みキーへフォールバックしない（無料ゲート＝ログイン必須）。
        # 配布版で未ログインなら key=None のまま False を返し、呼び出し側の REST
        # フォールバックへ落とす（そこで _dist_guard がログイン必須メッセージを出す）。
        logged_in = backend_client.is_logged_in()
        key = None if secrets.is_dist_build() else self._resolve_api_key()
        if not logged_in and not key:
            logger.warning("Deepgram 未ログイン（配布版）またはキー未設定のためストリーミングを開始できません")
            return False

        try:
            # websockets は遅延 import（未導入環境でもアプリ起動を阻害しない）
            from websockets.sync.client import connect  # noqa: F401
        except Exception as e:
            logger.warning(f"websockets が利用できないためストリーミングを無効化します: {e}")
            return False

        self._thread = threading.Thread(
            target=self._run, args=(key, logged_in), daemon=True, name="DeepgramStream"
        )
        self._thread.start()
        return True

    def send(self, samples: npt.NDArray[np.float32]) -> None:
        """
        16kHz モノラル Float32 チャンクを送信キューへ積む（audio スレッドから呼ばれる）。

        ネットワーク I/O もロック取得もせず、固定上限キューへノンブロッキング投入する
        だけなので、回線が遅くても audio コールバックは即座に return する。実送信は
        専用 sender スレッドが行う。キューが溢れたら音声を黙って捨てず streaming 失敗
        として記録し、保持済み全音声での REST フォールバックへ移す。
        """
        if samples is None or len(samples) == 0:
            return
        # フラグはプレーン読み（GIL 下で十分。ロックを取らないことで callback を塞がない）
        if self._cancelled or self._closed or self._failed:
            return
        pcm = _float32_to_pcm16le(samples)
        try:
            self._send_q.put_nowait(pcm)
        except queue.Full:
            # 送信が音声生成に追いつかない（回線が遅い/詰まり）。取りこぼしを隠さず
            # streaming 失敗にして REST フォールバック（task.audio_data の全音声）へ委ねる
            logger.warning("ストリーミング送信キューが上限に達しました。REST にフォールバックします")
            self._mark_failed()

    def finish(self, timeout: float = _FINISH_TIMEOUT) -> str:
        """
        送信を打ち切り、確定テキストを返す（ホットキーを離したときに呼ぶ）。

        CloseStream を送ると Deepgram は残りを確定して Metadata を返し接続を閉じる。
        取りこぼし対策で最大 timeout 秒だけ待つ。アプリのワーカースレッドから
        呼ばれるためブロックしてよい（UI/リスナーは塞がない）。

        Returns:
            確定テキスト（CJK 隣接スペース除去済み）。接続失敗・無音なら空文字
        """
        with self._lock:
            ws = self._ws

        # 接続確立前に呼ばれた場合、短時間だけ接続（または終了）を待つ。
        # これがないと短い発話で「接続前に finish → セッション破棄 → 空文字」になり、
        # 毎回 REST フォールバックへ落ちてストリーミングの利点が消える
        if ws is None:
            grace_end = time.monotonic() + _CONNECT_GRACE_SEC
            while time.monotonic() < grace_end and not self._done_event.is_set():
                with self._lock:
                    ws = self._ws
                    closed = self._closed
                if ws is not None or closed:
                    break
                time.sleep(0.02)

        with self._lock:
            ws = self._ws
            if ws is None:
                # 猶予内に接続できなかった → 受信スレッドに後始末させる
                self._cancelled = True

        if ws is not None:
            # CloseStream は送信キュー経由で積む（退避済み音声の後に送られ順序が保たれる）。
            # 詰まっていて積めない場合は最後の手段で直接送る（worker スレッドなのでブロック可）
            try:
                self._send_q.put_nowait(_CLOSE_STREAM)
            except queue.Full:
                try:
                    ws.send(json.dumps({"type": "CloseStream"}))
                except Exception as e:
                    logger.debug(f"CloseStream 送信エラー（無視）: {e}")

        # 確定（Metadata 受信）または切断まで待つ
        self._done_event.wait(timeout)
        self._shutdown_socket()
        if self._failed:
            # キュー溢れ/送信失敗 → 部分結果は使わず REST フォールバックに全音声で委ねる
            return ""
        return strip_cjk_spaces(self._current_text())

    def cancel(self) -> None:
        """結果を使わずに接続を破棄する（録音破棄時など）。"""
        with self._lock:
            self._cancelled = True
        self._shutdown_socket()
        self._resolve_finish()

    def _mark_failed(self) -> None:
        """ストリーミングを失敗として確定し、finish() を空文字で返させて REST に委ねる。

        キュー溢れ・送信失敗から呼ばれる。sender / 受信ループが closed を見て止まり、
        finish() の待機も解除されるので、呼び出し側は保持済み全音声で REST に落ちる。
        """
        with self._lock:
            if self._failed:
                return
            self._failed = True
            self._closed = True
        self._resolve_finish()

    def _shutdown_socket(self) -> None:
        """受信ループを止め、WebSocket を閉じる（多重呼び出し安全）。"""
        with self._lock:
            self._closed = True
            ws = self._ws
        if ws is not None:
            try:
                ws.close()
            except Exception as e:
                logger.debug(f"WebSocket close エラー（無視）: {e}")

    def _current_text(self) -> str:
        """確定 + 暫定を結合した現在の全文（スペース除去前）。"""
        with self._lock:
            parts = self._finals + ([self._interim] if self._interim else [])
        return "".join(parts)

    # ------------------------------------------------------------------
    # 受信スレッド
    # ------------------------------------------------------------------

    def _run(self, key: Optional[str], logged_in: bool) -> None:
        """接続 → sender スレッド起動 → 受信ループ（専用スレッド上）。

        Args:
            key: 直叩き用の Deepgram API キー（未ログイン時に使う。ログイン時は None でも可）
            logged_in: True なら自社サーバーから短命 JWT を取得して Bearer で接続する
        """
        from websockets.sync.client import connect

        # 製品版（ログイン済み）: サーバーから短命 JWT を取得し Bearer で接続。
        # 未ログイン: 従来どおり埋め込み/設定キーを Token で使う（並存ガード）。
        if logged_in:
            try:
                grant = backend_client.fetch_ephemeral_token()
                auth_header = f"Bearer {grant['token']}"
            except backend_client.BackendError as e:
                # 短命トークン取得失敗（未ログイン扱い・サブスク無効・通信失敗など）
                # → finish() を解決し、REST フォールバックに委ねる
                logger.warning(f"Deepgram 短命トークンの取得に失敗: {e}")
                self._resolve_finish()
                return
        else:
            auth_header = f"Token {key}"

        params = {
            "model": self.model,
            "language": self._stream_language,
            "encoding": "linear16",
            "sample_rate": "16000",
            "channels": "1",
            "interim_results": "true",
            "punctuate": "true",
            "smart_format": "true",
        }
        url = "wss://api.deepgram.com/v1/listen?" + urlencode(params)

        try:
            ws = connect(
                url,
                additional_headers={"Authorization": auth_header},
                open_timeout=_OPEN_TIMEOUT,
            )
        except Exception as e:
            # 接続失敗 → finish() を解決して REST フォールバックに委ねる
            logger.warning(f"Deepgram ストリーミング接続に失敗: {e}")
            self._resolve_finish()
            return

        logger.info(
            f"Deepgram ストリーミング開始 (model={self.model}, lang={self._stream_language})"
        )

        # cancel/close 済みなら即閉じる。問題なければ ws を公開して sender を起動する。
        # 接続前に send() がキューへ積んだ PCM も sender が FIFO で送るので順序は保たれる
        close_now = False
        with self._lock:
            if self._cancelled or self._closed:
                close_now = True
            else:
                self._ws = ws
        if close_now:
            try:
                ws.close()
            except Exception:
                pass
            self._resolve_finish()
            return

        # 送信専用スレッドを起動（ネットワーク I/O はここに集約し、audio コールバックは塞がない）
        self._sender = threading.Thread(
            target=self._sender_loop, args=(ws,), daemon=True, name="DeepgramSend"
        )
        self._sender.start()

        self._receive_loop(ws)
        # ループ脱出（切断・終端）→ finish() を解決
        self._resolve_finish()

    def _sender_loop(self, ws) -> None:
        """送信キューの PCM を WebSocket へ送る専用スレッド（ロックを持たずに I/O する）。

        audio コールバックは put_nowait するだけで、実際のネットワーク送信はここで行う。
        送信に失敗したら音声を黙って捨てず streaming 失敗として記録し REST フォール
        バックへ委ねる。CloseStream マーカーを受け取ったら送って終了する。
        """
        while True:
            if self._closed or self._failed:
                return
            try:
                item = self._send_q.get(timeout=_RECV_POLL_SEC)
            except queue.Empty:
                continue  # closed/failed の再確認のためループ継続
            if item is _CLOSE_STREAM:
                try:
                    ws.send(json.dumps({"type": "CloseStream"}))
                except Exception as e:
                    logger.debug(f"CloseStream 送信エラー（無視）: {e}")
                return  # CloseStream 後は送信不要（受信ループが Metadata を待つ）
            try:
                ws.send(item)  # ← ネットワーク I/O。_lock は保持していない
            except Exception as e:
                logger.debug(f"ストリーミング送信エラー: {e}")
                self._mark_failed()
                return

    def _receive_loop(self, ws) -> None:
        """確定/暫定メッセージを受信し続けるループ。"""
        # ConnectionClosed は websockets の例外。型に依存せずクラス名で握る
        while True:
            with self._lock:
                if self._closed:
                    return
            try:
                message = ws.recv(timeout=_RECV_POLL_SEC)
            except TimeoutError:
                continue  # closed フラグの再確認のためループ継続
            except Exception:
                # ConnectionClosed（正常終端含む）・その他切断
                return
            try:
                self._handle(message)
            except Exception as e:
                logger.debug(f"ストリーミング応答の処理でエラー（無視）: {e}")

    def _handle(self, message) -> None:
        """Deepgram のストリーミング応答 1 件を処理する。"""
        if isinstance(message, (bytes, bytearray)):
            message = message.decode("utf-8", errors="ignore")
        msg = json.loads(message)

        # Metadata は CloseStream 後の終端通知 → 確定完了とみなす
        if msg.get("type") == "Metadata":
            self._resolve_finish()
            return

        channels = (msg.get("channel") or {}).get("alternatives") or []
        if not channels:
            return
        transcript = (channels[0].get("transcript") or "").strip()

        with self._lock:
            if msg.get("is_final"):
                if transcript:
                    self._finals.append(transcript)
                self._interim = ""
            else:
                self._interim = transcript
            snapshot = "".join(self._finals + ([self._interim] if self._interim else []))

        # HUD へは日本語スペースを除去した見た目で渡す
        cb = self.on_interim
        if cb is not None:
            try:
                cb(strip_cjk_spaces(snapshot))
            except Exception:
                pass  # HUD 側の例外でストリーミングを止めない

    def _resolve_finish(self) -> None:
        """finish() の待機を一度だけ解除する（終端 / 切断 / タイムアウトのいずれか）。"""
        self._done_event.set()
